### ISSUE TITLE
Qwen3.5-122B-A10B MoE: output degenerates after ~200 tokens on MQ4

### ISSUE THREAD

**Issue by fivetide** [Kaden-Schutt/hipfire#145]
## Problem

The Qwen3.5-122B-A10B MoE model (`qwen3.5-122b-a10b.mq4`) produces coherent output for ~200 tokens, then degenerates into repeated punctuation/emoji garbage. Dense models (9B, 27B) work perfectly with the same engine.

**Reproducer:**
```bash
curl -s http://127.0.0.1:11435/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{"model":"qwen3.5-122b-a10b.mq4","messages":[{"role":"user","content":"how to center a div"}],"max_tokens":500}'
```

Output is coherent for ~200 tokens (CSS tutorial with Flexbox examples), then degenerates:
```
#### Center Horizontally Only (**Traditional**) :**):** )**):** ):** }:**| :** }):**| **:}):»** !** |:»:» !** !) **:** »» **«
```

**Key observations:**
- Deterministic — identical garbage across 3 runs
- Temperature-independent — temp=0 delays visible corruption but logits are wrong
- Short outputs (<200 tokens) work fine (e.g. counting to 50)
- Specific to MoE models, not dense

## Root Cause Analysis

Two compounding sources of accumulated quantization error:

### 1. DeltaNet Q8 state quantization error (~40% of drift)

The 122B MoE model operates at **10x smaller hidden-state magnitudes** than dense models (x_norm ≈ 20 vs ≈ 250 on 9B). This makes Q8 DeltaNet state quantization error proportionally larger. With 64 heads × 128×128 state matrices across 36 DeltaNet layers, the error compounds across tokens.

**Evidence:** FP32 DeltaNet state extends coherent output from ~200 to ~350 tokens. Cost: ~0.2 GB extra (negligible).

### 2. MQ4 expert weight quantization error (~60% of drift)

Each MoE layer applies 8 routed experts + 1 shared expert = 9 contributions to the residual stream. With 48 layers × 256 experts quantized to 4-bit, the accumulated quantization error through 9 expert contributions per layer compounds faster than a dense model's single FFN.

**Evidence:** Even with FP32 state, output degenerates at ~350 tokens into word fragments (`"soci eti es resi gn reali ti s"`).

### Hidden state trajectory

```
Early  (pos 14-50):  x_norm ≈ 14-21, x_max ≈ 2-8   — stable
Middle (pos 200-280): x_norm ≈ 19-22, x_max ≈ 4-8   — starting to drift
Late   (pos 285+):   x_norm ≈ 28-31, x_max ≈ 15-22  — exploding
```

### ✅ FP32 DeltaNet state (partial fix)
- Branch: `fix/moe-fp32-deltanet-state`
- Result: extends coherent output from ~200 to ~350 tokens
- Cost: ~0.2 GB extra VRAM
- Status: ready to merge

### ⚠️ MQ6 expert quantization (blocked by memory)
- Quantizer fixed to propagate `--format mq6` to expert weights (was hardcoded MQ4)
- Added disk spill to handle 90 GB output without OOM during quantization
- Successfully produced 90 GB MQ6 model file
- **Blocked:** 90 GB model doesn't load on 125 GB UMA (Strix Halo) — GTT + kernel overhead exceeds physical RAM
- Mixed quant (`--format mq4-mq6exp`: MQ4 attention + MQ6 experts ≈ 75 GB) couldn't quantize — 234 GB source safetensors mmaps + rayon temp buffers OOM

### ✅ HFQ loader `drop_mmap` optimization
- Makes `mmap` optional, dropped after header parsing on unix
- Prevents mmap'd source pages from competing with GPU allocations on UMA
- All tensor reads go through `pread` + `fadvise_dontneed`
- Status: ready to merge

## Recommended Path Forward

1. **Merge FP32 state fix** — immediate partial improvement, no downsides
2. **Merge loader + quantizer improvements** — enables MQ6 on systems with >128 GB RAM
3. **Future: streaming quantizer** — process safetensors files one at a time instead of mmap'ing all 39 simultaneously. Would allow mixed-quant on 125 GB systems
4. **Future: MQ5 format** — 5-bit would be ~72 GB for experts, fitting in 125 GB UMA

## Environment
- Hardware: AMD Strix Halo (gfx1151), 125 GB unified memory
- Model: Qwen3.5-122B-A10B (qwen3_5_moe, 48 layers, 256 experts, top-8)
- Quantization: MQ4G256 (hipfire-quantize)

**Comment by Kaden-Schutt** [comment]
## Reproducing on hipx (Strix Halo, gfx1151) post-#146 — fresh data

Confirming the analysis with a new repro on different hardware:

**Setup:**
- Hardware: Minisforum MS-S1 MAX, Ryzen AI MAX+ 395 (Strix Halo, gfx1151, 96 GB iGPU UMA pool, 32 GB OS).
- Branch: master @ \`ba31a90\` (#146 included) + JIT shim trio cherry-picked from \`strix-halo\`.
- Quant: fresh MQ4 from upstream \`Qwen/Qwen3.5-122B-A10B\` safetensors via \`hipfire-quantize --format mq4\` (65 GB output, 100% of params quantized excluding mtp/visual).

**Run with \`HIPFIRE_KV_MODE=asym3\`** (recommended default per \`reference_kv_quant_defaults.md\`):

\`\`\`
Prefill: 23 tokens / 1554ms = 15 tok/s
Decode:  2048 tokens / 56652ms = 36.2 tok/s    (71% of 256 GB/s LPDDR5x bandwidth ceiling)
No aperture violation. Run completes cleanly.
\`\`\`

**Quality observation matches #145's analysis exactly:**

The model produces ~50-100 coherent thinking-block tokens, then enters classic punctuation-loop degeneration:

\`\`\`
"Wait? Let me recall historical consensus on why Rome fell..."
\`\`\`

(repeating with progressively worse fragmentation)

After the \`</think>\` boundary, the answer body produces one coherent bullet point ("Political Instability: Frequent civil wars..."), then collapses into a digit-counting attractor for the rest of the output:

\`\`\`
2345678901234567890123456789012345678901234567890123456789012345678...
\`\`\`

**Two distinct attractor modes** observed in this single run:
1. Punctuation/word-fragment loop in the \`<think>\` block (#145's described mode)
2. Digit-counting loop in the answer body (new failure mode — appears to be triggered by markdown-list numbering "2." getting amplified by the same MQ4 expert error mechanism)

## Implications for the path forward

#146's FP32 DeltaNet state IS doing real work — we get more coherent tokens before degeneration than the pre-#146 ~200-token figure. But the MQ4 expert error remains dominant.

**Quality fix candidates worth ranking against this empirical baseline:**

1. **MQ6 expert (PR #147 path)**: 90 GB output. Fits hipx's 96 GB iGPU pool but very tight (no KV headroom). Mixed \`mq4-mq6exp\` (~75 GB) is what we'd actually want — currently blocked on the 234 GB safetensors mmap OOM during quantization (per #147 test plan).

2. **Streaming quantizer**: process safetensors files one at a time instead of mmap'ing all 39 simultaneously. Would unblock \`mq4-mq6exp\` on hipx (32 GB OS RAM) + would generalize to other big-model quants. Probably the single highest-leverage quantizer change.

3. **HFP4 / RDNA-native FP4 format**: storage-FP4 + compute-FP8 (or compute-FP16 on gfx1151). 4 bits/param storage = ~57 GB at 122B params. Better than MQ4 quality (per-block exponent preserves dynamic range) AND smaller. Multi-week design + PTQ work but is the "market upset" tier per the project vision.

## Related kernel bug filed

Filed #152 for the MoE \`gemv_hfq4g256_moe_gate_up_k8_indexed\` aperture violation observed on the **Q8 KV path** (not the asym3 path above). Likely triggered by degenerate-output routing values producing invalid expert indices. The Q8 path has known issues here; \`asym3\` is the working path for 122B-A10B.

**Comment by fivetide** [comment]
## Cross-reference: llama.cpp parity test confirms hipfire-specific degeneration

During investigation of #171, tested 122B-A10B MQ4 on a code-review prompt (md5 `d236222c364be82afcb9d745ddd8c467`):

| Prompt | Settings | Tokens | Unique words | Verdict |
|---|---|---|---|---|
| sheep (36pp) | temp=0.3, RP=1.05 | 42 (self-EOS) | high | **CLEAN** |
| code review (214pp) | temp=0.3, RP=1.05 | 1500 (max) | 14.2% | **ATTRACTOR** |
| code review (214pp) | greedy, RP=1.05 | 428 (self-EOS) | low | **ATTRACTOR** (repetition in output) |

Short prompts with low temperature are clean; longer outputs still degenerate. This is consistent with the original report of ~200-token coherence window.

For A3B models, llama.cpp Q4_K_M produces clean output on the same prompts where hipfire MQ4/HFQ4 hits attractors — confirming these are hipfire-specific, not model-intrinsic. Full parity test details in #171.

The fa0592d fix (split softmax+topk_renorm) fully resolved 3.5-A3B but 3.6-A3B and 122B remain affected. The remaining issue appears to be quantization-format precision on critical router weights (Q4_K_M's mixed-precision format preserves more precision than HFQ4G256 on weights near decision boundaries).

_Hardware: Strix Halo gfx1151, 128GB UMA._

**Comment by fivetide** [comment]
## Resolved: 122B-A10B clean after re-quantization with Q8 router (#180)

Re-quantized from original BF16 safetensors with the fix from PR #180 (unconditional Q8 router for MoE models). Tested on the same code-review prompt that previously triggered degeneration:

| Version | Tokens | Unique words | Verdict |
|---|---|---|---|
| Old MQ4 (4-bit router) | 428 (self-EOS) | 54.7% | Borderline (mild tail repetition) |
| Old MQ4, temp=0.3 | 1500 (max) | 14.2% | **ATTRACTOR** |
| **New MQ4 + Q8 router** | **487 (self-EOS)** | **59.9%** | **CLEAN** |

Root cause was the same as #171: 4-bit quantization of `mlp.gate.weight` (the MoE router) destroys routing precision. The 122B has 48 layers × 256 experts, so the compounding effect is even stronger than on the 35B-A3B models.

The fix (PR #180) makes Q8 router the default for all MoE models — cost is ~0.05% model size increase.

_Tested on Strix Halo gfx1151, 128GB UMA, asym3 KV mode. Prompt md5: `d236222c364be82afcb9d745ddd8c467`._
